Improve fnv64a hashing algorithm by remove all allocations

mateuszdyminski opened the pull request:
Hi everyone,

Lastly I've seen 2 great tutorial how to profile and optimize Go programs. Based on that knowledge I found some small improvements in your bigcache package. I started from adding allocation reporting for all of your benchmarks. After further analysis in pprof tool it showed me that your usage of fnv64a hash function was a bit inefficient. There were 2 not necessary allocations. Here is the output of your benchmarks before the changes:

> BenchmarkWriteToCacheWith1Shard-4                            	 2000000	       963 ns/op	     617 B/op	       5 allocs/op
> BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard-4     	 2000000	       828 ns/op	      65 B/op	       4 allocs/op
> BenchmarkWriteToUnlimitedCacheWithSmallInitSizeAnd1Shard-4   	 1000000	      1591 ns/op	    3470 B/op	       4 allocs/op
> BenchmarkWriteToCacheWith512Shards-4                         	 3000000	       540 ns/op	     611 B/op	       5 allocs/op
> BenchmarkWriteToCacheWith1024Shards-4                        	 3000000	       537 ns/op	     611 B/op	       5 allocs/op
> BenchmarkWriteToCacheWith8192Shards-4                        	 2000000	       581 ns/op	     628 B/op	       5 allocs/op
> BenchmarkWriteToCacheWith1024ShardsAndSmallShardInitSize-4   	 2000000	       642 ns/op	     823 B/op	       5 allocs/op
> BenchmarkReadFromCacheWith1024Shards-4                       	 5000000	       283 ns/op	      31 B/op	       4 allocs/op
> BenchmarkReadFromCacheWith8192Shards-4                       	 5000000	       283 ns/op	      31 B/op	       4 allocs/op
> PASS
> ok  	github.com/allegro/bigcache	28.775s

Benchmark results after my changes:

> BenchmarkWriteToCacheWith1Shard-4                            	 2000000	       886 ns/op	     577 B/op	       3 allocs/op
BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard-4     	 2000000	       761 ns/op	      41 B/op	       2 allocs/op
BenchmarkWriteToUnlimitedCacheWithSmallInitSizeAnd1Shard-4   	 1000000	      1545 ns/op	    3446 B/op	       2 allocs/op
BenchmarkWriteToCacheWith512Shards-4                         	 3000000	       495 ns/op	     571 B/op	       3 allocs/op
BenchmarkWriteToCacheWith1024Shards-4                        	 3000000	       501 ns/op	     571 B/op	       3 allocs/op
BenchmarkWriteToCacheWith8192Shards-4                        	 3000000	       533 ns/op	     590 B/op	       3 allocs/op
BenchmarkWriteToCacheWith1024ShardsAndSmallShardInitSize-4   	 3000000	       614 ns/op	     979 B/op	       3 allocs/op
BenchmarkReadFromCacheWith1024Shards-4                       	 5000000	       261 ns/op	      15 B/op	       2 allocs/op
BenchmarkReadFromCacheWith8192Shards-4                       	 5000000	       248 ns/op	      15 B/op	       2 allocs/op
PASS
ok  	github.com/allegro/bigcache	28.084s


And comparison taken from benchcmp tool(https://godoc.org/golang.org/x/tools/cmd/benchcmp):

> benchmark                                                      old ns/op     new ns/op     delta
BenchmarkWriteToCacheWith1Shard-4                              963           886           -8.00%
BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard-4       828           761           -8.09%
BenchmarkWriteToUnlimitedCacheWithSmallInitSizeAnd1Shard-4     1591          1545          -2.89%
BenchmarkWriteToCacheWith512Shards-4                           540           495           -8.33%
BenchmarkWriteToCacheWith1024Shards-4                          537           501           -6.70%
BenchmarkWriteToCacheWith8192Shards-4                          581           533           -8.26%
BenchmarkWriteToCacheWith1024ShardsAndSmallShardInitSize-4     642           614           -4.36%
BenchmarkReadFromCacheWith1024Shards-4                         283           261           -7.77%
BenchmarkReadFromCacheWith8192Shards-4                         283           248           -12.37%

> benchmark                                                      old allocs     new allocs     delta
BenchmarkWriteToCacheWith1Shard-4                              5              3              -40.00%
BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard-4       4              2              -50.00%
BenchmarkWriteToUnlimitedCacheWithSmallInitSizeAnd1Shard-4     4              2              -50.00%
BenchmarkWriteToCacheWith512Shards-4                           5              3              -40.00%
BenchmarkWriteToCacheWith1024Shards-4                          5              3              -40.00%
BenchmarkWriteToCacheWith8192Shards-4                          5              3              -40.00%
BenchmarkWriteToCacheWith1024ShardsAndSmallShardInitSize-4     5              3              -40.00%
BenchmarkReadFromCacheWith1024Shards-4                         4              2              -50.00%
BenchmarkReadFromCacheWith8192Shards-4                         4              2              -50.00%


I've tested it against Go1.6.2 and 1.7beta1. On both versions results were better between 2-12% in ns/op. Less number of allocations should also improve the GC pauses. 

The problem was that you have called fnv.New64a() every time. Under the hood it allocates bytes and it could be omit. One solution might be sync.Pool but it's not reliable in case of GC. Another is just take the Go standard library fnv64a implementation and remove the allocation which is made in New64a func. 

Inspiration for this PR:
https://youtu.be/ZuQcbqYK0BY 
https://youtu.be/N3PWzBeLX2M

janisz commented on bigcache_bench_test.go:
```diff
@@ -15,6 +15,7 @@ func BenchmarkWriteToCacheWith1Shard(b *testing.B) {
 }
 
 func BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard(b *testing.B) {
+	b.ReportAllocs()
```
I'd prefer to use `-test.benchmem` on whole test suite.

janisz commented on hash.go:
```diff
@@ -18,8 +14,17 @@ func newDefaultHasher() Hasher {
 type fnv64a struct {
```
Maybe we should move this structure and it's functions and constants to separate file and add tests for it. Also adding dockstring comments and pointing to original source is necessary.

janisz commented:
Hi @mateuszdyminski 

Thanks for PR. I left comments on it.

mateuszdyminski commented:
Moved implementation to fnv.go file.

Created test and benchmark for new fnv hash implementation. Here is the output from benchmark:
> BenchmarkFnvHashSum64-4             100000000       11.1 ns/op     0 B/op         0 allocs/op
> BenchmarkFnvHashStdLibSum64-4   20000000         67.3 ns/op   16 B/op         2 allocs/op

I also removed all ReportAllocs(). Thanks for tip with -benchmem flag.

janisz commented:
LGTM!
@druminski would you like to review this patch?

druminski commented:
@mateuszdyminski @janisz I have just analyzed this PR and I like it very much, great job and thanks for the contribution :)

janisz commented:
We need to generate new benchmarks results for [README](https://github.com/allegro/bigcache#benchmarks)

druminski commented:
Ok, new bench results generated and updated in [readme](https://github.com/allegro/bigcache#benchmarks)

mateuszdyminski commented:
Nice! Thanks for accepting PR : )
